proposals/adaptive-do-notation.rst

```diff
@@ -0,0 +1,114 @@
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/127>`_.
+.. sectnum::
+.. contents::
+
+The ``Prelude.>>=`` combinator limits the return type as a ``m b``, which cause Monads not composable. In order to ease this restrictions, this proposal introduces another type class ``Dsl`` for do notation.
+
+Motivation
+----------
+
+Monads do not compose. Normally a ``do`` block cannot contains operators defined in different monad instances.
+
+The state of art solution is `using only one monad <http://okmij.org/ftp/Haskell/extensible/>`_ , ``Eff``, which forwards all monadic bind operations to custom effect handlers, instead of defining new monads.
+
+However, the ``Eff`` approach is heavy weight than ordinary monad. It's not very convenient to create an additional indirect layer for simple use cases.
+
+The ``Eff`` approach of bypassing ``>>=`` combinator is quite embarrassing. Since ``>>=`` settles on our logo, In this proposal we present a new approach to enable multiple operations at once by improving ``>>=``. This proposal aims to port the approach used in `Dsl.scala <https://github.com/ThoughtWorksInc/Dsl.scala>`_ to GHC. This approach improves the extensibility of `>>=`.
+
+Proposed Change Specification
+-----------------------------
+
+=======================
+Changes in the compiler
+=======================
+
+Desugaring ``<-`` to ``Dsl.>>=`` instead of ``Prelude.>>=``.
+
+======================
+Changes in the library
+======================
+
+----------------------------
+Add a new type class ``Dsl``
+----------------------------
+
+::
+
+  module Dsl where
+  {-# LANGUAGE FlexibleInstances #-}
+  {-# LANGUAGE MultiParamTypeClasses #-}
+  class Dsl m a d where
+    (>>=) :: m a -> (a -> d) -> d
+    (>>) :: m a -> d -> d
+    (>>) ma d = ma Dsl.>>= \a -> d
+
+The ``>>=`` combinator of the ``Dsl`` type class is similar to ``Prelude.>>=``, except ``m b`` is replaced to ``d``. As a result, multiple ``<-`` expressions in a ``do`` block can have the same ``d`` and different ``m``.
+
+
+----------------------------
+Add a forwarder to ``Monad``
+----------------------------
+
+::
+
+  instance Monad m => Dsl m a (m b) where
+    (>>=) = (Prelude.>>=)
+
+Effect and Interactions
+-----------------------
+
+Suppose you want to create a random number generator. The generated numbers should be stored in a lazily evaluated infinite list, which can be built with the help of adaptive do notation.
+
+::
+
+  import Data.Int
+  import Data.List
+  import Data.Bits
+
+  newtype Yield a = Yield { element :: a }
+
+  instance Dsl Yield a [a] where
+    interpret (Yield a) handler = a : handler a
```

Comment:
I'm confused. I don't see a definition of `>>=`, and I don't see a class method called `interpret`.